nntrainer/layers/layer_internal.h

```diff
@@ -454,24 +466,54 @@ class Layer : public ml::train::Layer {
    */
   virtual int initialize(Manager &manager) = 0;
 
+  /**
+   * @brief get number of input layers
+   */
   virtual unsigned int getNumInputs() { return input_dim.size(); }
 
+  /**
+   * @brief get number of output layers
+   */
   virtual unsigned int getNumOutputs() { return output_dim.size(); }
 
+  /**
+   * @brief set Number of Input Layers
+   */
   void setNumInputs(unsigned int size) {
     if (size < 1)
       throw std::invalid_argument("Minimum number of inputs must be 1");
     input_dim.resize(size);
     net_input.resize(size);
   }
 
+  /**
+   * @brief set Number of Output Layers
+   */
   void setNumOutputs(unsigned int size) {
     if (size < 1)
       throw std::invalid_argument("Minimum number of outputs must be 1");
     output_dim.resize(size);
     net_hidden.resize(size);
   }
 
+  /**
+   * @brief Set the input dimension
+   * @param[in] d dimension to be set
+   */
+  void setInputDimension(const std::vector<TensorDim> &d) { input_dim = d; }
+
+  /**
+   * @brief Set the input dimension
+   * @param[in] d dimension to be set
+   * @param[in] i axis
+   */
+  void setInputDimension(const TensorDim &d, unsigned int i) {
```

Comment:
is it necessary to be make these public? can you please use `setProperty` as these will have to removed eventually?

Reply:
> is it necessary to be make these public? can you please use `setProperty` as these will have to removed eventually?

as you know, there are no interface to get the input dimension except input layer. The input dimension and output dimension is calculated inside of layer just before initialize() is called. So.. setProperty state is quite earlier than this and also it is not proper to ask more information from user.  Is there any reason not to be public ??

Reply:
input dimension for a layer can be updated using `setProperty` just like any property.
`input_shape` is already supported as a basic property for all layers. 

> So.. setProperty state is quite earlier than this and also it is not proper to ask more information from user

`setProperty` still works for this, allowing to update the `input_shape`. As internal code will be doing it, it can be assumed that it will be done properly. Please consider updating this in a later PR.

> Is there any reason not to be public ??

Opening this up means that all custom layers will have to provide support for another function call.